### PR TITLE
address some clang-tidy issues

### PR DESCRIPTION
**Pull request by zingale**
mostly narrowing and nullptrs

## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
quick tests: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-02-25-001/index.html

File: Source/driver/Castro.cpp [hunk 1]
```diff
@@ -1438,7 +1439,7 @@ Castro::init (AmrLevel &old)
 {
     BL_PROFILE("Castro::init(old)");
 
-    Castro* oldlev = (Castro*) &old;
+    Castro* oldlev = dynamic_cast<Castro*>(&old);
```
Comment (yut23) [inline Source/driver/Castro.cpp:1442]: Do all the compilers we care about support this? cppreference says some implementations use RTTI

Comment (zingale) [reply Source/driver/Castro.cpp:1442]: not sure.  I can remove that.

Comment (zingale) [reply Source/driver/Castro.cpp]: reverted

File: Source/driver/Castro.cpp [hunk 2]
```diff
@@ -2710,7 +2713,7 @@ Castro::reflux(int crse_level, int fine_level)
 
             // Start with a MultiFab version of the flux register.
 
-            for (OrientationIter fi; fi; ++fi) {
+            for (OrientationIter fi; fi != nullptr; ++fi) {
```
Comment (yut23) [inline Source/driver/Castro.cpp:2716]: I think `fi.isValid()` is better here

File: Source/driver/Castro_io.cpp [hunk 3]
```diff
@@ -994,10 +994,10 @@ Castro::plotFileOutput(const std::string& dir,
...
         //
-        for (int i =0; i < static_cast<int>(plot_var_map.size()); i++)
+        for (auto& p : plot_var_map)
         {
-            int typ = plot_var_map[i].first;
-            int comp = plot_var_map[i].second;
+            int typ = p.first;
+            int comp = p.second;
             os << desc_lst[typ].name(comp) << '\n';
         }
```
Comment (yut23) [inline Source/driver/Castro_io.cpp:1002]: We can use a structured binding declaration here:
```
        for (const auto& [typ, comp] : plot_var_map) {
            os << desc_lst[typ].name(comp) << '\n';
        }
```

**Review by yut23 — CHANGES_REQUESTED** [review]
Looks good aside from my comment on Castro.cpp

**Review by yut23 — APPROVED** [review]

**Pull request merged**